local_pathfinding/global_path.py

```diff
@@ -0,0 +1,247 @@
+        # but this would be a breaking change to multiple files, so this is a quick workaround
+
+        # we don't need to update multiple times per second
+        time.sleep(1)
+
+
+def get_most_recent_file(directory_path: str) -> str:
+    """
+    Returns the most recently modified file in the specified directory.
+
+    Args:
+        directory_path (str): The path to the directory containing the files.
+
+    Returns:
+        str: The path to the most recently modified file.
+    """
+    all_files = os.listdir(directory_path)
+
+    # Filter out directories and get the full file paths
+    files = [
+        os.path.join(directory_path, file)
+        for file in all_files
+        if os.path.isfile(os.path.join(directory_path, file))
+    ]
+
+    # Sort the files based on their last modification time
+    files.sort(key=lambda x: os.path.getmtime(x), reverse=True)
+
+    if files:
+        return files[0]
+    else:
+        return ""
+
+
+def get_path(file_path: str) -> Path:
+    """Returns the global path from the specified file path.
+
+    Args:
+        file_path (str): The path to the global path csv file.
+
+    Returns:
+        Path: The global path retrieved from the csv file.
+    """
+    path = Path()
+
+    with open(file_path, "r") as file:
+        reader = csv.reader(file)
+        # skip header
+        reader.__next__()
+        for row in reader:
+            path.waypoints.append(HelperLatLon(latitude=float(row[0]), longitude=float(row[1])))
+    return path
+
+
+def post_path(path: Path) -> bool:
+    """Sends the global path to NET via POST request.
+
+    Args:
+        path (Path): The global path.
+
+    Returns:
+        bool: Whether or not the global path was successfully posted.
+    """
+    waypoints = [
+        {"lat": float(item.latitude), "lon": float(item.longitude)} for item in path.waypoints
+    ]
+    data = {"waypoints": waypoints}
+    json_data = json.dumps(data).encode("utf-8")
+    try:
+        urlopen(PATH_URL, json_data)
+        return True
+    except HTTPError as http_error:
+        print(f"HTTP Error: {http_error.code}")
+    except URLError as url_error:
+        print(f"URL Error: {url_error.reason}")
+    except Exception as e:
+        print(f"An unexpected error occurred: {e}")
+
+    return False
+
+
+def get_pos() -> HelperLatLon:
+    """Returns the current position of sailbot, retrieved from the an http GET request.
+
+    Returns:
+        HelperLatLon: The current position of sailbot.
+    """
+    try:
+        position = json.loads(urlopen("http://localhost:3005/api/gps").read())
+    except HTTPError as http_error:
+        print(f"HTTP Error: {http_error.code}")
+    except URLError as url_error:
+        print(f"URL Error: {url_error.reason}")
+    except Exception as e:
+        print(f"An unexpected error occurred: {e}")
+
+    lat_lon = position["lat_lon"]
+    pos = HelperLatLon(latitude=lat_lon["latitude"], longitude=lat_lon["longitude"])
+
+    return pos
```

Comment:
@SPDonaghy if you run the website (uncomment docker-compose.website.yml in devcontainer.json then rebuild dev container) you will get a response from the endpoint.

However, currently it is empty: 
```
{'success': True, 'data': []}
```

@UBCSailbot/website can you populate the GPS endpoint with something?

Reply:
@patrick-5546  Sure! I was also thinking if we don't get a response or if its empty we would just print to console and keep trying the GET request? Since the path we generate with a constant wont be too meaningful anyways.

Reply:
Here's an example of what the GPS endpoint returns when populated:

```
{
	success: true,
	data: [
		{
			latitude: 49.376629216076275,
			longitude: -123.27266500802254,
			speed: 0,
			heading: 0,
			timestamp: "2024-01-19T20:42:50.509Z"
		},
		{
			latitude: 49.37711663428484,
			longitude: -123.27156381625609,
			speed: 0,
			heading: 0,
			timestamp: "2024-01-19T20:42:50.510Z"
		}
	]
}
``` 

There's also the option to populate the endpoint by running our tests locally:

1. Access the website container terminal on Docker
2. Go into the directory "tests", e.g. `cd tests` 
3. Run `npm install`
4. Run `npm run test-tag @gps` 

This should write some sample gps data into DB, which you can access via the endpoint.

Reply:
I ran the website by uncommenting the docker-compose.website.yml file and rebuilding the container, but I am having trouble populating the endpoint with data because npm is not a recognized command. Is there something else I need to change to make sure npm is installed?

Reply:
You have to run it in the website container from docker desktop

![image](https://github.com/UBCSailbot/local_pathfinding/assets/65756895/fe62264e-1cfd-484d-9335-6d7afabc5a02)